oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,284 @@
+        val output = modelService.simByName[request.name]?.output
+        if (output != null) {
+            output.offer(request)
+        } else {
+            throw IllegalStateException("no simulation '${request.name}' or buffer full")
+        }
+        SimulationResultConfirmDTO.newBuilder().build()
+    }
+
+    override fun offerErrorResult(request: ErrorResponseDTO, responseObserver: StreamObserver<ErrorConfirmDTO>) = responseObserver.consume {
+        val error = modelService.simByName[request.name]?.error
+        if (error != null) {
+            error.offer(request)
+        } else {
+            throw IllegalStateException("no simulation '${request.name}' or buffer full")
+        }
+        ErrorConfirmDTO.newBuilder().build()
+    }
+
+    override fun offerSimulationConfig(request: NodeStatusCommandOrResponseDTO, responseObserver: StreamObserver<NodeChangeConfirmDTO>) = responseObserver.consume {
+        val update = modelService.simByName[request.name]?.update
+
+        if (update != null) {
+            update.offer(request)
+        } else {
+            throw IllegalStateException("no simulation '${request.name}' or buffer full")
+        }
+        NodeChangeConfirmDTO.newBuilder().build()
+    }
+
+    override fun updateNode(request: NodeStatusCommandOrResponseDTO, responseObserver: StreamObserver<NodeChangeConfirmDTO>) = responseObserver.consume {
+        val newNode = updateFromResponse(request)
+        modelService.updateSim(request.name!!, newNode)
+        modelService.updateStatusMessage("${request.name} updated")
+        NodeChangeConfirmDTO.newBuilder().setMessage("Simulation updated with inputs: ${newNode.inputs} outputs: ${newNode.outputs}").build()
+    }
+
+
+    override fun sendMessage(request: MessageCommandDTO, responseObserver: StreamObserver<MessageReponseDTO>) = responseObserver.consume {
+        println("Message from [${request.name}] : ${request.message}")
+        modelService.addMessage(Message(request.name, request.message))
+        MessageReponseDTO.newBuilder().build()
+    }
+
+    override fun unregister(request: UnRegistrationRequestDTO, responseObserver: StreamObserver<UnRegistrationResponseDTO>) = responseObserver.consume {
+        var unregistered = false
+        if (request.name in modelService.simByName.keys) {
+            val name = request.name
+            modelService.removeSim(name)
+            unregistered = true
+        }
+        UnRegistrationResponseDTO.newBuilder().setUnregistered(unregistered).build()
+    }
+
+    private fun updateFromResponse(request: NodeStatusCommandOrResponseDTO): Simulation {
+        return modelService.simByName.getValue(request.name).copy(
+                inputs = request.inputsList.map { Input(it.name, it.lowerBound, it.upperBound, it.currentValue) },
+                outputs = request.outputsList.map { Output(it.name) },
+                description = request.description
```

Comment:
hmmm. This simulation object it gives back, it contains a copy of all the data but the same channels.

This kind of mostly-identity-but-sometimes-reference equality really gives me the willies.